### PR TITLE
feat: x/oracle simulations

### PR DESCRIPTION
**Pull request by adamewozniak**
Creates oracle simulations 

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/oracle/module.go [hunk 1]
```diff
@@ -169,3 +173,54 @@ func (am AppModule) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.Val
 
 	return []abci.ValidatorUpdate{}
 }
+
+// AppModuleSim implements the AppModule interface for the x/oracle module.
+type AppModuleSim struct {
```
Comment (adamewozniak) [inline x/oracle/module.go:178]: Had to add the `accountKeeper` and `bankKeeper` for weighted operations but didn't want to touch `AppModule` - is this the appropriate way to do this? Or should we just use `AppModule` and add the `accountKeeper` and `bankKeeper` to it 
cc : @alexanderbez

Comment (toteki) [reply x/oracle/module.go:178]: Answer to this will be relevant to @RafilxTenfen 's PR

Comment (alexanderbez) [reply x/oracle/module.go:178]: So yes, typically you add these keeper dependencies in `AppModule`, which you then pass those types to the relevant simulation logic/calls.

That being said, you can also define a special `AppModuleSim`, but AFAIK, that isn't the standard in the SDK. So if we're going with consistency, you can just have them set on `AppModule`. But I don't have a strong preference either way. I just wanted be consistent _within umee_.

Comment (alexanderbez) [reply x/oracle/module.go:178]: ref: https://github.com/umee-network/umee/pull/401#discussion_r787990311

File: x/oracle/simulations/decoder.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+	"github.com/umee-network/umee/x/oracle/types"
+)
+
+// NewDecodeStore returns a decoder function closure that unmarshals the KVPair's
+// Value to the corresponding oracle type.
+func NewDecodeStore(cdc codec.Codec) func(kvA, kvB kv.Pair) string {
+	return func(kvA, kvB kv.Pair) string {
+		switch {
+		case bytes.Equal(kvA.Key[:1], types.KeyPrefixExchangeRate):
```
Comment (alexanderbez) [inline x/oracle/simulations/decoder.go:21]: nit: I personally prefer putting a newline after each `case` body. It makes it slightly easier to read IMO.

File: x/oracle/simulations/genesis.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	votePeriodKey               = "vote_period"
+	voteThresholdKey            = "vote_threshold"
+	rewardBandKey               = "reward_band"
+	rewardDistributionWindowKey = "reward_distribution_window"
+	slashFractionKey            = "slash_fraction"
+	slashWindowKey              = "slash_window"
+	minValidPerWindowKey        = "min_valid_per_window"
+)
+
+// GenVotePeriod randomized VotePeriod
```
Comment (alexanderbez) [inline x/oracle/simulations/genesis.go:24]: Can we add to the godocs for each of these of what value ranges they generate?

e.g. `// GenVotePeriod produces a randomized VotePeriod in the range of [x,y]`

File: x/oracle/simulations/genesis.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	rewardBandKey               = "reward_band"
+	rewardDistributionWindowKey = "reward_distribution_window"
+	slashFractionKey            = "slash_fraction"
+	slashWindowKey              = "slash_window"
+	minValidPerWindowKey        = "min_valid_per_window"
+)
+
+// GenVotePeriod produces a randomized VotePeriod in the range of [1, 100]
+func GenVotePeriod(r *rand.Rand) uint64 {
+	return uint64(1 + r.Intn(100))
```
Comment (alexanderbez) [inline x/oracle/simulations/genesis.go:26]: I think we should do this within [5, 100]. WDYT?

File: x/oracle/simulations/operations.go [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+	return strings.Join(exchangeRates, ",")
+}
+
+// WeightedOperations returns all the operations from the module with their respective weights
+func WeightedOperations(
+	appParams simtypes.AppParams,
+	cdc codec.JSONCodec,
+	ak types.AccountKeeper,
+	bk types.BankKeeper,
+	k keeper.Keeper) simulation.WeightedOperations {
```
Comment (alexanderbez) [inline x/oracle/simulations/operations.go:59]:
```suggestion
	k keeper.Keeper,
) simulation.WeightedOperations {
```

File: x/oracle/simulations/operations.go [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+	var (
+		weightMsgAggregateExchangeRatePrevote int
+		weightMsgAggregateExchangeRateVote    int
+		weightMsgDelegateFeedConsent          int
+	)
+	appParams.GetOrGenerate(cdc, OpWeightMsgAggregateExchangeRatePrevote, &weightMsgAggregateExchangeRatePrevote, nil,
```
Comment (alexanderbez) [inline x/oracle/simulations/operations.go:65]:
```suggestion

	appParams.GetOrGenerate(cdc, OpWeightMsgAggregateExchangeRatePrevote, &weightMsgAggregateExchangeRatePrevote, nil,
```

File: x/oracle/simulations/operations.go [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+	ak types.AccountKeeper,
+	bk types.BankKeeper,
+	k keeper.Keeper,
+) simtypes.Operation {
+	return func(
+		r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simtypes.Account, chainID string,
+	) (simtypes.OperationMsg, []simtypes.FutureOperation, error) {
+		simAccount, _ := simtypes.RandomAcc(r, accs)
+		address := sdk.ValAddress(simAccount.Address)
+		// ensure the validator exists
```
Comment (alexanderbez) [inline x/oracle/simulations/operations.go:110]:
```suggestion

		// ensure the validator exists
```

File: x/oracle/simulations/operations.go [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+			return simtypes.NoOpMsg(
+				types.ModuleName,
+				types.TypeMsgAggregateExchangeRatePrevote,
+				"unable to find validator",
+			), nil, nil
+		}
+
+		prices := make(map[string]sdk.Dec, len(acceptList))
+		for _, denom := range acceptList {
+			prices[denom] = simtypes.RandomDecAmount(r, sdk.NewDec(1)).Add(umeePrice)
```
Comment (alexanderbez) [inline x/oracle/simulations/operations.go:122]: we should use `r` to generate a random value [0,1] and add that to `umeePrice` IMO.

File: x/oracle/simulations/operations.go [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+	OpWeightMsgAggregateExchangeRateVote    = "op_weight_msg_exchange_rate_aggregate_vote"
+	OpWeightMsgDelegateFeedConsent          = "op_weight_msg_exchange_feed_consent"
+
+	salt = "1234"
+)
+
+var (
+	acceptList                    = []string{types.UmeeSymbol, types.USDDenom}
+	voteHashMap map[string]string = make(map[string]string)
+	umeePrice                     = sdk.NewDec(25)
```
Comment (alexanderbez) [inline x/oracle/simulations/operations.go:33]: Can we use something like `25.71` here? `sdk.MustNewDecFromStr("25.71")`

File: x/oracle/simulations/operations.go [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+		if err != nil {
+			return simtypes.NoOpMsg(types.ModuleName, msg.Type(), "unable to generate mock tx"), nil, err
+		}
+
+		_, _, err = app.Deliver(txGen.TxEncoder(), tx)
+		if err != nil {
+			return simtypes.NoOpMsg(types.ModuleName, msg.Type(), "unable to deliver tx"), nil, err
+		}
+
+		voteHashMap[address.String()] = exchangeRatesStr
```
Comment (alexanderbez) [inline x/oracle/simulations/operations.go:165]: I see why this `voteHashMap` is needed, but are we sure it's safe for use between simulation iterations?

Comment (adamewozniak) [reply x/oracle/simulations/operations.go:165]: We can definitely put this in an instance of a struct or something

Comment (adamewozniak) [reply x/oracle/simulations/operations.go:169]: Localized it to `WeightedOperations`, let me know if that's alright

Comment (alexanderbez) [reply x/oracle/simulations/operations.go:169]: I think that should work 👍

File: x/oracle/simulations/operations.go [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+			return simtypes.NoOpMsg(
+				types.ModuleName,
+				types.TypeMsgAggregateExchangeRatePrevote,
+				"unable to find validator",
+			), nil, nil
+		}
+
+		prices := make(map[string]sdk.Dec, len(acceptList))
+		for _, denom := range acceptList {
+			prices[denom] = umeePrice.Add(simtypes.RandomDecAmount(r, sdk.NewDec(1)))
```
Comment (alexanderbez) [inline x/oracle/simulations/operations.go:126]: Aren't we supposed to generate a random value between 0.0 and 1.0 here?

**Review by alexanderbez — APPROVED** [review]

**Pull request merged**